daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala

```diff
@@ -65,7 +65,27 @@ object Speedy {
     def popEnv(count: Int): Unit =
       env.subList(env.size - count, env.size).clear
 
-    /* Compute a stack trace from the locations in the continuation stack. */
+    /** Push a single location to the continuation stack for the sake of
+        maintaining a stack trace. */
+    def pushLocation(loc: Location): Unit = {
+      lastLocation = Some(loc)
+      val last_index = kont.size() - 1
+      val last_kont = if (last_index >= 0) Some(kont.get(last_index)) else None
+      last_kont match {
+        case Some(KArg(Array(SEValue(SToken)))) => kont.add(last_index, KLocation(loc))
```

Comment:
Some comments on the first two cases would be nice.

Reply:
Done. I'm in the middle of simplifying this a bit. The explanation will (hopefully) get a bit simpler as a result of this.